### PR TITLE
Split learner with composition.

### PR DESCRIPTION
**Pull request by trivialfis**
Another version of  #5350, use composition instead of inheritance as suggested by @RAMitchell .  Keeping the old branch for comparison.

### CODE REVIEW

File: src/learner.cc [hunk 1]
```diff
@@ -193,60 +190,129 @@ void GenericParameter::ConfigureGpuId(bool require_gpu) {
...
-  explicit LearnerImpl(std::vector<std::shared_ptr<DMatrix> > cache)
-      : need_configuration_{true} {
-    monitor_.Init("Learner");
-    for (std::shared_ptr<DMatrix> const& d : cache) {
-      cache_.Cache(d, GenericParameter::kCpuId);
+struct LearnerAttributes {
+  /*! \brief Constant string identifying a passed in parameter is a metric name. */
+  static std::string const kEvalMetric;  // NOLINT
+  /*! \brief A global storage for internal prediction. */
+  PredictionContainer cache;
```
Comment (RAMitchell) [inline src/learner.cc]: If any of these variables are only used by one of the learner classes and do not need to be shared, move them to be members of that class. The idea is that variables are not shared if they do not need to be.

File: src/learner.cc [hunk 2]
```diff
@@ -193,60 +190,129 @@ void GenericParameter::ConfigureGpuId(bool require_gpu) {
...
-      omp_set_num_threads(generic_parameters_.nthread);
+    if (attr->generic_parameters.nthread != 0) {
+      omp_set_num_threads(attr->generic_parameters.nthread);
     }
 
     // add additional parameters
     // These are cosntraints that need to be satisfied.
-    if (tparam_.dsplit == DataSplitMode::kAuto && rabit::IsDistributed()) {
-      tparam_.dsplit = DataSplitMode::kRow;
+    if (attr->tparam.dsplit == DataSplitMode::kAuto && rabit::IsDistributed()) {
```
Comment (RAMitchell) [inline src/learner.cc]: Are we able to remove data split mode in the near future?

File: src/learner.cc [hunk 3]
```diff
@@ -345,141 +481,148 @@ class LearnerImpl : public Learner {
...
 
-  void CheckDataSplitMode() {
-    if (rabit::IsDistributed()) {
-      CHECK(tparam_.dsplit != DataSplitMode::kAuto)
-        << "Precondition violated; dsplit cannot be 'auto' in distributed mode";
-      if (tparam_.dsplit == DataSplitMode::kCol) {
-        // 'distcol' updater hidden until it becomes functional again
-        // See discussion at https://github.com/dmlc/xgboost/issues/1832
-        LOG(FATAL) << "Column-wise data split is currently not supported.";
+  void ConfigureNumFeatures(LearnerAttributes* attr) {
```
Comment (RAMitchell) [inline src/learner.cc]: Can we remove this in the near future and use DMatrix as the source of truth?

File: src/learner.cc [hunk 4]
```diff
@@ -345,141 +481,148 @@ class LearnerImpl : public Learner {
...
+    CHECK_NE(attr->mparam.num_feature, 0)
+        << "0 feature is supplied.  Are you using raw Booster interface?";
+    // Remove these once binary IO is gone.
+    attr->cfg["num_feature"] = common::ToString(attr->mparam.num_feature);
+    attr->cfg["num_class"] = common::ToString(attr->mparam.num_class);
+  }
+
+  void ConfigureGBM(LearnerAttributes *attr,
+                    LearnerTrainParam const &old, Args const &args) const {
+    if (attr->gbm == nullptr || old.booster != attr->tparam.booster) {
```
Comment (RAMitchell) [inline src/learner.cc]: If Learner had access to parameters on its construction, we could get rid of these chains of Configure methods.

File: src/learner.cc [hunk 5]
```diff
@@ -572,44 +715,43 @@ class LearnerImpl : public Learner {
...
+    LearnerModelParamLegacy mparam = attr->mparam;  // make a copy to potentially modify
     std::vector<std::pair<std::string, std::string> > extra_attr;
     mparam.contain_extra_attrs = 1;
 
     {
       std::vector<std::string> saved_params;
       // check if rabit_bootstrap_cache were set to non zero before adding to checkpoint
-      if (cfg_.find("rabit_bootstrap_cache") != cfg_.end() &&
-        (cfg_.find("rabit_bootstrap_cache"))->second != "0") {
+      if (attr->cfg.find("rabit_bootstrap_cache") != attr->cfg.end() &&
```
Comment (RAMitchell) [inline src/learner.cc]: Not sure what `rabit_bootstrap_cache` is

File: src/learner.cc (hunk 2 above)
Comment (trivialfis) [reply src/learner.cc]: Depends.  Do we still want to support different split mode in the future?

File: src/learner.cc (hunk 3 above)
Comment (trivialfis) [reply src/learner.cc]: Once the JVM PR is merged.

File: src/learner.cc (hunk 4 above)
Comment (trivialfis) [reply src/learner.cc]: It doesn't.  The parameters can be loaded from pickle.

File: src/learner.cc (hunk 5 above)
Comment (trivialfis) [reply src/learner.cc]: It's from JVM package.

**Comment by trivialfis** [comment]
I'm afraid the variables are indeed shared.

**Comment by trivialfis** [comment]
@RAMitchell I prefer https://github.com/dmlc/xgboost/pull/5350 over this PR.  I know that jumping around inheritance is not ideal, but at least that way we don't need an extra `this` pointer (`attr` in this PR).  The split PR is only to split up the functionality so we can remove them as much as possible in the future.

**Comment by RAMitchell** [comment]
Sounds good to me.

**Pull request closed without merging**